### PR TITLE
Topbar search submit

### PR DESCRIPTION
**Pull request by kpuputti**
Construct the search query from the existing URL params and whatever the SearchBar component returns.

Required params:

 - locale: `locale`
 - view type: `view` (map, list, grid)
 - keyword query: `q`
 - location query: `lq`
 - coordinates: `lc`
 - places service status: `ls`
 - bounding box viewport for the map: `boundingbox`
 - max search distance: `distance_max`

### CODE REVIEW

File: app/view_utils/topbar_helper.rb [hunk 1]
```diff
@@ -54,10 +57,15 @@ def topbar_props(community:, path_after_locale_change:, user: nil, search_placeh
...
-        location_placeholder: 'Location'
+        mode: main_search.to_s,
+        keyword_placeholder: 'Search...', # TODO: localise
+        location_placeholder: 'Location', # TODO: localise
+
+        # TODO: figure where to get these
+        # keyword_query: params[:q],
+        # location_query: params[:lq]
       },
+      search_path: '/', # TODO: fix
```
Comment (kpuputti) [inline app/view_utils/topbar_helper.rb]: This could be passed within the search object.

File: client/app/components/sections/Topbar/Topbar.js [hunk 2]
```diff
@@ -1,5 +1,8 @@
+/* eslint-disable no-console */
```
Comment (kpuputti) [inline client/app/components/sections/Topbar/Topbar.js]: This is temporary and should be removed when the logging is removed.

File: client/app/components/sections/Topbar/Topbar.js [hunk 3]
```diff
@@ -41,7 +44,7 @@ const avatarDropdownProps = (avatarDropdown, customColor, username, routes) => {
 const LABEL_TYPE_MENU = 'menu';
 const LABEL_TYPE_DROPDOWN = 'dropdown';
 
-const SEARCH_ENABLED = false;
+const SEARCH_ENABLED = true;
```
Comment (kpuputti) [inline client/app/components/sections/Topbar/Topbar.js]: This  whole flag should be removed when the search works.

File: client/app/components/sections/Topbar/Topbar.js [hunk 4]
```diff
@@ -92,6 +95,55 @@ const DEFAULT_CONTEXT = {
   loggedInUsername: null,
 };
 
+const parseQuery = (searchQuery) => {
```
Comment (kpuputti) [inline client/app/components/sections/Topbar/Topbar.js]: See if the [URL](https://developer.mozilla.org/en/docs/Web/API/URL) API could be used for parsing.

File: client/app/startup/TopbarApp.js [hunk 5]
```diff
@@ -1,5 +1,6 @@
 import r from 'r-dom';
 import { initialize as initializeI18n } from '../utils/i18n';
+
 import Topbar from '../components/sections/Topbar/Topbar';
 import { subset } from '../utils/routes';
```
Comment (kpuputti) [inline client/app/startup/TopbarApp.js]: TopbarApp should be made to use the new TopbarContainer to get the actions.

File: client/app/components/sections/Topbar/Topbar.js [hunk 6]
```diff
@@ -121,6 +121,41 @@ const DEFAULT_CONTEXT = {
   loggedInUsername: null,
 };
 
+const currentSearchParams = (searchQuery) => {
+  const PARAMS_TO_KEEP = ['view', 'locale'];
```
Comment (kpuputti) [inline client/app/components/sections/Topbar/Topbar.js]: Are these the only params we want to keep form the current URL?

File: client/app/components/composites/MenuPriority/MenuPriority.css [hunk 7]
```diff
@@ -26,4 +26,7 @@
   }
 }
 
-.hiddenLinks {} /* stylelint-disable-line block-no-empty */
+.hiddenLinks {
+  /* Fixes dropdown being below old filter controls and admin panel selectize inputs. */
+  z-index: 2;
```
Comment (Gnito) [inline client/app/components/composites/MenuPriority/MenuPriority.css]: How about other dropdowns? Language (another Menu) & AvatarDropdown

File: client/app/components/composites/SearchBar/SearchBar.css [hunk 8]
```diff
@@ -390,4 +322,28 @@
...
+  position: absolute;
+  bottom: -1px;
+  right: calc(var(--SearchBar_mobileHeight) / 2 - var(--SearchBar_mobileTriangleSize));
+  border: var(--SearchBar_mobileTriangleSize) solid transparent;
+
+  /* We must adjust the brightness here to match either the input
+     background or the submit button background when the toggle
+     position changes in the breakpoint. This cannot be done in JS by
+     switching the color since the breakpoints are unknown there. */
+  filter: brightness(80%);
```
Comment (Gnito) [inline client/app/components/composites/SearchBar/SearchBar.css]: Filter brightness doesn't work well with IE11 & Edge. 
Since this arrow is invisible most of the time, so as an alternative, we could just create 2 arrows (with custom colors from client/app/utils/colors.js) and show them based on media queries?

Comment (Gnito) [inline client/app/components/composites/SearchBar/SearchBar.css]: It seems we already calculate this brightness(80%) in component

File: client/app/components/sections/Topbar/Topbar.js [hunk 9]
```diff
@@ -233,18 +263,31 @@ class Topbar extends Component {
...
+          keywordQuery: oldSearchParams.q,
+          locationQuery: oldSearchParams.lq,
+          customColor: marketplace_color1,
+          onSubmit: ({ keywordQuery, locationQuery, place }) => {
+            const query = createQuery({
+              q: keywordQuery,
+              lq: locationQuery,
+              lc: placesUtils.coordinates(place),
+              boundingbox: placesUtils.viewport(place),
+              distance_max: placesUtils.maxDistance(place),
```
Comment (Gnito) [inline client/app/components/sections/Topbar/Topbar.js]: One additional parameter to consider is service status 
(we used that to figure out Google maps related API problems like quota etc.)
However, it can be added after dl if needed.

File: client/app/components/composites/SearchBar/SearchBar.css (hunk 8 above)
Comment (kpuputti) [inline client/app/components/composites/SearchBar/SearchBar.css]: You are correct. This CSS side is older and the brightness calculation in the component is a more recent addition. I'll set the color in the JS side.

Comment (kpuputti) [inline client/app/components/composites/SearchBar/SearchBar.css]: No wait, I cannot really do this. The reason why the filter is in the CSS is that it changes based on the viewport width.

On a narrow screen the arrow is next to the darker submit button:

screen shot 2016-08-09 at 10 14 50

but on a wider screen the arrow is next to the lighter input background:

screen shot 2016-08-09 at 10 18 17

And since JS isn't aware of the breakpoints we cannot change the color dynamically without a resize handler that checks the current breakpoint. And I'd like to avoid doing that.

Comment (kpuputti) [inline client/app/components/composites/SearchBar/SearchBar.css]: One possible solution would be to create two separate arrows that have bg color assigned in JS, and the CSS would just swap between those in the breakpoint.

File: client/app/components/composites/MenuPriority/MenuPriority.css (hunk 7 above)
Comment (Gnito) [inline client/app/components/composites/MenuPriority/MenuPriority.css]: I'll make another Trello todo task from this

File: client/app/components/sections/Topbar/Topbar.js (hunk 9 above)
Comment (kpuputti) [inline client/app/components/sections/Topbar/Topbar.js]: Done.

File: client/app/components/composites/MenuPriority/MenuPriority.css (hunk 7 above)
Comment (kpuputti) [inline client/app/components/composites/MenuPriority/MenuPriority.css]: Let's do this in another PR.

File: client/app/components/composites/SearchBar/SearchBar.css (hunk 8 above)
Comment (kpuputti) [inline client/app/components/composites/SearchBar/SearchBar.css]: Let's postpone this since the colors are a bit open anyways.

File: app/controllers/homepage_controller.rb [hunk 10]
```diff
@@ -286,28 +297,31 @@ def search_coordinates(latlng)
...
       @current_community.private &&
       !@current_user
   end
 
+  def search_modes_in_use(q, lc, main_search)
+    [
+      lc && (main_search == :location || main_search == :keyword_and_location),
+      q && (main_search == :keyword || main_search == :keyword_and_location)
+    ]
+  end
```
Comment (lyyder) [inline app/controllers/homepage_controller.rb]: Does this return an array of booleans which indicates what search modes are in use? Would it be clearer if the search mode order was not swapped (param order vs. return value)?

File: app/controllers/homepage_controller.rb [hunk 11]
```diff
@@ -57,34 +55,24 @@ def index
...
-        sw_lat, sw_lng, ne_lat, ne_lng = coords
-        { boundingbox: { sw: [sw_lat, sw_lng], ne: [ne_lat, ne_lng] } }
-      elsif params[:lc].present?
-        { center: params[:lc].split(',') }
-      else
-        Maybe(@current_community.location)
-          .map { |l| { center: [l.latitude, l.longitude] }}
-          .or_else(nil)
-      end
+      viewport = viewport_geography(params[:boundingbox], params[:lc], @current_community.location)
```
Comment (mporkola) [inline app/controllers/homepage_controller.rb]: I think `viewport_geometry` is a better term here

**Comment by lyyder** [comment]
👍  for the js utility libs and the API docs on those.

File: client/app/components/sections/Topbar/Topbar.css [hunk 12]
```diff
@@ -54,16 +54,29 @@
     }
   }
 
+  & button {
```
Comment (mporkola) [inline client/app/components/sections/Topbar/Topbar.css]: Since we've ended up using tag selector (I trust it was a well-considered choice), I'd like to make it as specific as possible. So maybe a child selector (`& > button`) instead of a more broader descendant selector?

File: client/app/components/sections/Topbar/Topbar.css [hunk 13]
```diff
@@ -54,16 +54,29 @@
...
+    border-radius: 0;
+  }
+
   & .topbarLogo {
     max-height: var(--Topbar_logoMaxHeightMobile);
-    max-width: var(--Topbar_textLogoMaxWidth);
+    max-width: 70%;
```
Comment (mporkola) [inline client/app/components/sections/Topbar/Topbar.css]: To me this looks like the `max-width` is 70% of the whole top bar container, that's probably wrong?

File: client/app/components/sections/Topbar/Topbar.js [hunk 14]
```diff
@@ -121,6 +121,36 @@ const DEFAULT_CONTEXT = {
   loggedInUsername: null,
 };
 
+const SEARCH_PARAMS_TO_KEEP = ['view', 'locale'];
+const parseKeepParams = urlUtils.currySearchParams(SEARCH_PARAMS_TO_KEEP);
+const SEARCH_PARAMS = ['q', 'lq'];
+const parseSearchParams = urlUtils.currySearchParams(SEARCH_PARAMS);
+
+const isValid = (value) => typeof value === 'number' && !isNaN(value) || !!value;
```
Comment (mporkola) [inline client/app/components/sections/Topbar/Topbar.js]: maybe more information about what kind of "valid" we're talking about? e.g. `isValidSearchParam`

File: client/app/utils/places.js [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+
+const toRadians = (degrees) => degrees * (Math.PI / 180); // eslint-disable-line no-magic-numbers
+
+/**
+ * Calculate the distance between two points.
+ *
+ * @param {google.maps.LatLng} pointA - first point
+ * @param {google.maps.LatLng} pointB - second point
+ *
+ * @return {Number} distance between the given points (in km?)
```
Comment (mporkola) [inline client/app/utils/places.js]: Yes, in km :) This comes from the unit used in `EARTH_RADIUS` constant below: if we want rods or furlongs we can just change the radius to desired unit.

File: client/app/utils/places.js [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+ */
+const computeScale = (pointA, pointB) => {
```
Comment (mporkola) [inline client/app/utils/places.js]: The naming of this function was my failure, it should be fixed :) It actually doesn't do what the document promises (I think?), but instead returns `distance / 2`. I suggest this should be changed to `getDistance` and then the `/ 2` be done elsewhere.

File: app/controllers/homepage_controller.rb (hunk 10 above)
Comment (Gnito) [inline app/controllers/homepage_controller.rb]: :+1:
We thought that destructuring was possible with Hashes, but ruby didn't like it so had to change it to array - and we didn't notice this one

File: client/app/components/sections/Topbar/Topbar.css (hunk 12 above)
Comment (Gnito) [inline client/app/components/sections/Topbar/Topbar.css]: :+1: or explicit comment that this is override of R&A styles

File: app/controllers/homepage_controller.rb (hunk 10 above)
Comment (kpuputti) [inline app/controllers/homepage_controller.rb]: Good point. This was originally a hash that was attempted to destructure, but was changed to an array. I'll change this.

File: client/app/utils/places.js (hunk 16 above)
Comment (Gnito) [inline client/app/utils/places.js]: Howabout getBoundingboxRadius? We don't need distance between 2 points anywhere atm. (ie. splitting can be done later too if necessary)

File: client/app/components/sections/Topbar/Topbar.css (hunk 12 above)
Comment (kpuputti) [inline client/app/components/sections/Topbar/Topbar.css]: Those generic selectors within the Topbar.css file can be thought of being our CSS normalisation from the old styles.

That specific rule undoes the rounded corners that we have in all buttons in old styles, so I think it makes sense to keep quite generic and get the same effect to all future buttons in the Topbar.

Does this make sense to you?

File: client/app/components/sections/Topbar/Topbar.css (hunk 13 above)
Comment (kpuputti) [inline client/app/components/sections/Topbar/Topbar.css]: That's exactly what it is, and it's on purpose.

The max-width is for the text logo in the mobile breakpoint where we want to fit as much text in the Topbar as we can.

File: client/app/components/sections/Topbar/Topbar.css (hunk 12 above)
Comment (mporkola) [inline client/app/components/sections/Topbar/Topbar.css]: Yes, makes sense. Just for clarity, would it be possible to make a `resetStylesLeakingFromRails.css` (naming left as exercise) and include it to root components?

File: client/app/utils/places.js (hunk 16 above)
Comment (mporkola) [inline client/app/utils/places.js]: I think radius isn't a correct term, and in any case "distance" is exactly what we're calculating here. The change is very minor, I can do it myself as well :)

File: client/app/components/sections/Topbar/Topbar.css (hunk 12 above)
Comment (kpuputti) [inline client/app/components/sections/Topbar/Topbar.css]: @mporkola Great idea!
